File: HW1/Codes/22002717_Iynem_Arda.lua
function main()
  local operationNumber = 1

  -- Operation 1: Declare/create an empty list
  print("\nOperation " .. operationNumber)
  local languages = {}
  printResultingList(languages, operationNumber)
  operationNumber = operationNumber + 1

  -- Operation 2: Initialize a list with some values
  print("\nOperation " .. operationNumber)
  languages = { "Dart", "Go", "Javascript", "Lua", "Python", "Ruby", "Rust" }
  printResultingList(languages, operationNumber)
  operationNumber = operationNumber + 1

  -- Operation 3: Check if the list is empty or not
  print("\nOperation " .. operationNumber)
  isListEmpty(languages)
  printResultingList(languages, operationNumber)
  operationNumber = operationNumber + 1

  -- Operation 4: Add a new element to a list
  print("\nOperation " .. operationNumber)
  local newElement = "SQL"
  table.insert(languages, newElement)
  printResultingList(languages, operationNumber)
  operationNumber = operationNumber + 1

  -- Operation 5: Check if a particular element exists in the list
  print("\nOperation " .. operationNumber)
  doesContain(languages, newElement)
  printResultingList(languages, operationNumber)
  operationNumber = operationNumber + 1

  -- Operation 6: Remove a particular element from the list
  print("\nOperation " .. operationNumber)
  removeFromList(languages, newElement)
  printResultingList(languages, operationNumber)
  operationNumber = operationNumber + 1

  -- Operation 7: Get the head and the tail of a list
  print("\nOperation " .. operationNumber)
  local first = languages[1]
  local tail = { unpack(languages, 2) }
  print("Head of the list is: " .. first)
  print("Tail of the list is: " .. table.concat(tail, " "))
  printResultingList(languages, operationNumber)
  operationNumber = operationNumber + 1

  -- Operation 8: Print all of the elements in the list
  print("\nOperation " .. operationNumber)
  printResultingList(languages, operationNumber)
  operationNumber = operationNumber + 1
end

function isListEmpty(list)
  if next(list) == nil then
    print("List is empty")
  else
    print("List is not empty")
  end
end

function doesContain(list, object)
  for _, value in pairs(list) do
    if value == object then
      print("List contains " .. object)
      return
    end
  end
  print("List does not contain " .. object)
end

function removeFromList(list, object)
  for index, value in pairs(list) do
    if value == object then
      table.remove(list, index)
      return
    end
  end
end

function printResultingList(list, operationNumber)
  print("List after Operation " .. operationNumber .. ": " .. table.concat(list, " "))
end

main()

File: HW1/Codes/22002717_Iynem_Arda.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	operationNumber := 1

  // Operation 1: Declare/create an empty list
	fmt.Println(fmt.Sprintf("\nOperation %v", operationNumber))
	languages := []string{}
	printResultingList(languages, operationNumber)
	operationNumber = operationNumber + 1
  
  // Operation 2: Initialize a list with some values
	fmt.Println(fmt.Sprintf("\nOperation %v", operationNumber))
	languages = []string{"Dart", "Go", "Javascript", "Lua", "Python", "Ruby", "Rust"}
	printResultingList(languages, operationNumber)
	operationNumber = operationNumber + 1

  // Operation 3: Check if the list is empty or not
	fmt.Println(fmt.Sprintf("\nOperation %v", operationNumber))
	isListEmpty(languages)
	printResultingList(languages, operationNumber)
	operationNumber = operationNumber + 1
  
  // Operation 4: Add a new element to a list
	fmt.Println(fmt.Sprintf("\nOperation %v", operationNumber))
	newElement := "SQL"
	languages = append(languages, newElement)
	printResultingList(languages, operationNumber)
	operationNumber = operationNumber + 1
  
  // Operation 5: Check if a particular element exists in the list
	fmt.Println(fmt.Sprintf("\nOperation %v", operationNumber))
	doesContain(languages, newElement)
	printResultingList(languages, operationNumber)
	operationNumber = operationNumber + 1
  
  // Operation 6: Remove a particular element from the list
	fmt.Println(fmt.Sprintf("\nOperation %v", operationNumber))
	languages = removeFromList(languages, newElement)
	printResultingList(languages, operationNumber)
	operationNumber = operationNumber + 1
  
  // Operation 7: Get the head and the tail of a list
	fmt.Println(fmt.Sprintf("\nOperation %v", operationNumber))
	first := languages[0]
	last := languages[1:len(languages)]
	fmt.Println("Head of the list is: " + first)
	fmt.Println("Tail of the list is: " + strings.Join(last, " "))
	printResultingList(languages, operationNumber)
	operationNumber = operationNumber + 1
  
  // Operation 8: Print all of the elements in the list
	fmt.Println(fmt.Sprintf("\nOperation %v", operationNumber))
	printResultingList(languages, operationNumber)
	operationNumber = operationNumber + 1
}

func isListEmpty(list []string) {
	if len(list) > 0 {
		fmt.Println("List is not empty")
	} else {
		fmt.Println("List is empty")
	}
}

func doesContain(list []string, object string) int {
	for i, value := range list {
		if value == object {
			fmt.Println("List contains " + object)
			return i
		}
	}

	fmt.Println("List does not contain " + object)
	return 0
}

func removeFromList(list []string, object string) []string {
	for index, value := range list {
		if value == object {
			return append(list[:index], list[index+1:]...)
		}
	}
	return list
}

func printResultingList(list []string, operationNumber int) {
	fmt.Println(fmt.Sprintf("List after Operation %v: %s", operationNumber, strings.Join(list, " ")))
}